File: examples/03-interactions/python/start_retry_pattern.py
from temporalio.client import Client
from temporalio.exceptions import WorkflowAlreadyStartedError


# If the client times out or loses connectivity while starting a Workflow, the
# result is ambiguous: Temporal may have committed WorkflowExecutionStarted but
# the response never reached the caller. Retry with the same Workflow ID and
# treat "already started" as success.
async def start_order(client: Client, order_id: str) -> None:
    workflow_id = f"order-{order_id}"

    try:
        await client.start_workflow(
            OrderWorkflow.run,
            order_id,
            id=workflow_id,
            task_queue="orders",
        )
        print(f"Started {workflow_id}")
    except WorkflowAlreadyStartedError:
        print(f"Already started; treating retry as success: {workflow_id}")

File: examples/03-interactions/go/start_retry_pattern.go
package interactions

import (
	"context"
	"fmt"

	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/temporal"
)

// If the client times out or loses connectivity while starting a Workflow, the
// result is ambiguous: Temporal may have committed WorkflowExecutionStarted but
// the response never reached the caller. Retry with the same Workflow ID and
// treat "already started" as success.
func StartOrder(ctx context.Context, c client.Client, orderID string) error {
	workflowID := "order-" + orderID

	_, err := c.ExecuteWorkflow(
		ctx,
		client.StartWorkflowOptions{
			ID:        workflowID,
			TaskQueue: "orders",
		},
		OrderWorkflow,
		orderID,
	)
	if temporal.IsWorkflowExecutionAlreadyStartedError(err) {
		fmt.Println("Already started; treating retry as success:", workflowID)
		return nil
	}
	if err != nil {
		return err
	}

	fmt.Println("Started", workflowID)
	return nil
}

